### DISCUSSION TITLE
Midi output for more controls?

### DISCUSSION THREAD

**Discussion by maks** [bangcorrupt/hacktribe#136] [Ideas]
So I'm working on [a "tracker style" companion app for the E2 for editing patterns and instrument patches](https://github.com/maks/elfer).

Its very early stages at the moment, but one of my goals is for the app to be controlled (as in user input) from the **E2 itself**. 

At the moment per the factory firmware midi implementation I get midi messages for most of the "panel controls" so I'm able to use the pads for selecting parts as well steps, but there are no messages for the "step buttons" (1,2,3,4), Shift, the arrow buttons or the Part-arrow buttons. Ideally would also be great to have access to all those and the  OSC dial (because its nicely stepped) and the menu/enter, exit & write buttons 🙂 

But just having access to the steps, arrows and shift would make it so much easier for me to build a great set of controls for my apps.

@bangcorrupt I'm hoping this is perhaps something that is not too difficult to add with the current hacktribe firmware, as you have already done intercepting receiving midi messages for the purpose of poking/writing ram addresses?

**Comment by maks** [comment]
I also want to add my thanks to @rafamj who's "elecmidi" C header I used (and [documented how with Dart in this article](https://medium.com/@mksl/a-quick-trick-to-make-binary-data-in-flutter-a-breeze-a467fe1ac7da)) as well as porting some of the encoding/decoding functions in my Dart code.

**Reply by rafamj** [reply]
Hello,

  In the header some fields are defined as filler<0>, because I didn't use them. 

 To whoever writes something new for the Electribe: besides the step edit, in my view another shortcoming of the Electribe is that the motion sequence can only be assigned to values on the Panel Control (Ej: Osc Pitch, Osg Glide....) but there is not a way of assign it to an arbitrary CC value. This would be very useful when using the Electribe as a sequencer to control external synths.

**Reply by bangcorrupt** [reply]
> I had a motion sequence outputting as MIDI CC at one point but it needs a lot more work before it's useable (if it ever is). Motion sequences are interpolated 96 ticks per step, which caused a crash.

**Reply by maks** [reply]
Actually I'm rather confused about motion seuencing, though I'm trying to use it in the "normal" way in my app, eg to motion sequence a Ifx, osc pitch, etc.

From reading the korg midi impl doc, it says that there are 64bytes of motion sequencing data per sequence "slot", so I had thoght this was across all the steps of a part, so if a part was a full 4 bars (64 steps) long, then the motion sequence could only set 1 value per step, but I think I may be getting this wrong somehow?

**Reply by bangcorrupt** [reply]
You're not wrong, one byte per step regardless of pattern length.  In 'Smooth' mode, values are interpolated 96 ticks per step.  Each of the 24 motion sequences has a target channel and target parameter.  Some notes [here](https://pastebin.com/raw/Dsu5L5uw) and [there](https://pastebin.com/raw/t6bLPGmn).

**Reply by maks** [reply]
Thanks again! I had seen the stepped vs smooth settings for motion sequencing previously, now it makes perfect sense. 👍🏻

**Comment by bangcorrupt** [comment]
I haven't tried yet, but if I can I'll add CC messages for every physical control.  Maybe even have the pads send different messages depending on which mode is selected.

**Reply by maks** [reply]
Thank you! that would be fantastic to have all the controls send MIDI messages! And even having different ones from the pads in different modes would really be icing on the cake!

**Reply by maks** [reply]
@bangcorrupt I'm getting closer to having an initial working version of my tracker app (https://github.com/maks/elfer). 
For now I've only using computer keyboard keys for controlling the app (I'm using x/shift/arrow keys copying somewhat the same keys as the Dirtywave M8 tracker uses) and was thinking that I could use some of the pads (in trigger mode) for now for those same keys too, but while that works, the downside to that is that they trigger sounds of those parts if you don't have the volume down.

Not trying to rush you at all, but was wondering if you had a sense of how hard it will be to change the firmware for sending midi mesgs from the shift/arrow/1-4 buttons/select-dial would be?

**Reply by bangcorrupt** [reply]
I started on this last night.  I think I can add NRPN for all the buttons, hopefully with pads and bar buttons sending different values per pad mode.  I'm nowhere near testing anything, but I can see where to hijack execution and what variables are needed.

**Reply by bangcorrupt** [reply]
Initial support for button CC output added in fa9c04e.

**Reply by maks** [reply]
@bangcorrupt very excited to try this out! but sorry I think I maybe doing something wrong. 
I cloned the repo, checked out the `cc-buttons` branch then did:
```
sha256sum -c hash/SYSTEM.VSB.sha
SYSTEM.VSB: OK

bspatch4 SYSTEM.VSB hacked-SYSTEM.VSB patch/hacktribe-2.patch

sha256sum -c hash/hacked-SYSTEM.VSB.sha
hacked-SYSTEM.VSB: OK

python scripts/e2-header.py hacked-SYSTEM.VSB
```
Then with my blue e2 running an older version of the Hacktribe firmware I formattted a SDCard then placed the hacked-SYSTEM.VSB into `KORG\electribe sampler\System` and renamed to `SYSTEM.VSB`

But when I tried to do the Software Update menu option on the E2 I get "Error Invalid File".
Is this because I don't need to do the header hack because the Hacktribe firmware already makes the E2 think its a sampler so I don't need to do the header hack step?

Sorry to be so pedantic, just want to overly cautious and make sure I'm not making a wrong assumption and then muck up the firmware update.

**Reply by bangcorrupt** [reply]
> Is this because I don't need to do the header hack because the Hacktribe firmware already makes the E2 think its a sampler so I don't need to do the header hack step?

Yes, hacktribe is sampler firmware. There is also a [patching script](https://github.com/bangcorrupt/hacktribe/wiki/How-To#patch-automatically) to automate those commands.

EDIT:  The most recent patch will only see files under the `hacktribe` directory.  Copy the hacked VSB to:

`KORG/hacktribe/System/SYSTEM.VSB`

if you already installed [79a4e4d](https://github.com/bangcorrupt/hacktribe/commit/79a4e4d41eacdbebcb5dba1393a6d595ae9d5e8d).

**Reply by maks** [reply]
Phew, thanks for the update, yes I did install the first commit and then got the "Error no file" but then I spotted your edit above and I've now got the latest firmware installed with the encoder support 🎉 

My thanks again @bangcorrupt !! 🙏🏻  It all works perfectly! I can see the NRPN messages (I had to read up on what nrpn's are 🙂 so all makes sense now) all coming thru, even for the main select encoder by itself when I have the E2's config set to pattern-change only with shift+encoder, which is *perfect* for using it to control my app.
This really made my day, thanks! 👍🏻  I've started to use the buttons & encoders now to control my app, I'll post a demo video soon.

**Comment by maks** [comment]
@bangcorrupt I finally had a chance to record a quick demo (well sorry its actually a bit rambling) of the current state of my "companion app" for the E2 https://www.youtube.com/watch?v=-vGEu_1CHNk

Again huge thanks for implementing the full controls Midi NRPN messages in Hacktribe! without that I don't think my goal of controlling the app completely from the E2 would ever be possible.

**Reply by bangcorrupt** [reply]
At some point I'll add NRPN for the trigger pads in each mode as well, maybe the XY Pad too.

**Reply by maks** [reply]
Now I've got all the basics working, having trigger pads in other modes would really be icing on the cake! 👍🏻 
Eg. having pads in sequencer mode would let me use them to turn on/off a step in elfer to match what happens on the E2. Same for things like part mute.
Having access to step jump would be especially nice too, as I would just hijack that completely for something like bulk select/edit operations that trackers have, as its not even really used outside of playback performances.
But I've already got plenty to work on now in the meantime thanks to all the new controls your latest firmware has given me 🙂

**Reply by bangcorrupt** [reply]
> step jump would be especially nice too, as I would just hijack that completely

I may be able to create a 'local control off' mode where all of the controls only send MIDI, not affecting the internal state.

**Reply by maks** [reply]
Wow, being able to have a 'local control off' would be awesome! I didn't dare ask for it before as I thought that it would be too difficult to implement.
Having that would indeed be **very** helpful for me, as I've started experimenting with controlling other gear assigned to Parts using the "AudioIn Osc" via Midi and for instance currently I'm using the Value/Select dial to change instruments rather than the more obvious `Oscillator` dial because it of course would change the actual oscillator assigned to the part on the E2. I'll post a video of this experiment soon.
It would also help with a bunch of other slight annoyances I've run into with my experiments so having this would be all round great!

**Reply by bangcorrupt** [reply]
MIDI functions are relatively easy to implement now.  I think a message to request resetting the internal saved CC state is probably needed too, so that the next control change sends a full NRPN instead of just the data byte.  I'm not sure if the standard 'reset all controllers' will do this without changing the internal parameter settings.

**Reply by maks** [reply]
Yes having a message to request a full NRPN would be quite handy as I've run into this already: on connecting my app doesn't know which dial is the currently selected one, so I often have to twiddle another dial then the one I actually want in order to get that "current control" state into my app.

**Reply by bangcorrupt** [reply]
Try sending CC 0x79 (Reset All Controllers) to all parts.  I think the value is ignored.

**Reply by maks** [reply]
Cool, thanks!
I've given that a go, but I think I might be doing something wrong. I'm sending `[0xB0, 0x79, 0]` once I'm connected to the E2 but that doesn't seem change that behaviour, the previously selected dial doesn't send the full NRPN message. Or would I need to be using the latest dev branch firmware for that CC to work?

**Reply by bangcorrupt** [reply]
That looks right.  It's part of the factory firmware, but maybe doesn't work how I think it does.  I'll have a proper look tomorrow.

**Reply by bangcorrupt** [reply]
Turns out MIDI Reset All Controllers isn't supposed to reset _all_ controllers ([documentation](https://www.midi.org/component/edocman/response-to-reset-all-controllers/fdocument?Itemid=9999)).  I'll add a sysex function to reset the saved NRPN state.  I fixed a bug with the encoders, so it should now send a full correct NRPN the first time after boot.

**Reply by maks** [reply]
Brilliant, thanks again!!
Just having it sent after boot covers pretty much all I need. If you get a chance to do the reset state sysex that would be great, but for me having the "disconnect local control" for the controls that you mentioned in the other thread would be more handy for me.

**Comment by bangcorrupt** [comment]
In the next update, NRPN map will change slightly.  Everything will shift right by one byte, so what is currently:

[ NRPN-MSB, NRPN-LSB, DATA-MSB ] 

will become:

[ NRPN-LSB, DATA-MSB, DATA-LSB ] 

with NRPN-MSB used to select different modes (like fx editing).  Only the CC numbers will change, the values remain the same.  For panel control output, NRPN-MSB == 0.

**Reply by maks** [reply]
No worries at all, thanks for the heads up 👍🏻 
So if I understand correctly, its now possible to tell that eg "IFX On" is enabled, ie. the button is lit up? 
I'll grab the latest dev branch commit and try it out.

**Reply by bangcorrupt** [reply]
No, not yet.  You would still have to scrape this from the pattern file.  I added initial support for editing the FX parameters to create new presets, currently undocumented and incomplete.
